examples/bridge-app/linux/include/Clusters.h

```diff
@@ -0,0 +1,276 @@
+    virtual EmberAfStatus Write(const EmberAfAttributeMetadata * attributeMetadata, uint8_t * buffer)                        = 0;
+    virtual void WriteFromBridge(chip::AttributeId attributeId, const uint8_t * buffer)                                      = 0;
+    virtual chip::Span<const EmberAfAttributeMetadata> GetAllAttributes()                                                    = 0;
+};
+
+template <typename T, size_t Bytes, EmberAfAttributeType AfType>
+struct PrimitiveType
+{
+    static constexpr EmberAfAttributeType kMatterType = AfType;
+    // Note that Bytes can be < sizeof(T) due to non-power-of-two byte lengths.
+    static constexpr uint16_t kMaxSize = Bytes;
+    static_assert(Bytes <= sizeof(T), "Incorrect type / byte specification");
+    typedef T ValueType;
+
+    PrimitiveType() = default;
+    PrimitiveType(T initial) : mValue(initial) {}
+
+    EmberAfStatus Read(uint8_t * buffer, uint16_t sz) const
+    {
+        if (sz != kMaxSize)
+            return EMBER_ZCL_STATUS_FAILURE;
+        memcpy(buffer, &mValue, kMaxSize);
+        return EMBER_ZCL_STATUS_SUCCESS;
+    }
+
+    bool Equal(const uint8_t * buffer) const { return !memcmp(buffer, &mValue, kMaxSize); }
+
+    void Write(const uint8_t * buffer) { memcpy(&mValue, buffer, kMaxSize); }
+
+    bool TryWrite(ValueType newV)
+    {
+        if (mValue == newV)
+            return false;
+        mValue = newV;
+        return true;
+    }
+    bool TryWrite(const uint8_t * buffer)
+    {
+        if (Equal(buffer))
+            return false;
+        Write(buffer);
+        return true;
+    }
+
+    const ValueType & Peek() const { return mValue; }
+
+    uint8_t * GetBytes() { return reinterpret_cast<uint8_t *>(&mValue); }
+
+private:
+    ValueType mValue;
+};
+
+template <typename T>
+using StructType = PrimitiveType<T, sizeof(T), ZCL_STRUCT_ATTRIBUTE_TYPE>;
+
+template <size_t Bytes, EmberAfAttributeType AfType>
+struct OctetString
+{
+    static constexpr EmberAfAttributeType kMatterType = AfType;
+    static constexpr bool kIsLong =
+        (AfType == ZCL_LONG_CHAR_STRING_ATTRIBUTE_TYPE || AfType == ZCL_LONG_OCTET_STRING_ATTRIBUTE_TYPE);
+    static constexpr uint16_t kMaxSize = Bytes;
+    typedef const uint8_t * ValueType;
+
+    OctetString() = default;
+
+    EmberAfStatus Read(uint8_t * buffer, uint16_t sz) const
+    {
+        // There should be at least enough space to hold the size
+        if (kIsLong && sz < 2)
+            return EMBER_ZCL_STATUS_FAILURE;
+        memcpy(buffer, mValue, sz < kMaxSize ? sz : kMaxSize);
+        return EMBER_ZCL_STATUS_SUCCESS;
+    }
+
+    bool Equal(const uint8_t * buffer)
+    {
+        uint16_t newsz = GetSize(buffer);
+        if (newsz != GetSize(mValue))
+            return false;
+        return memcmp(buffer, mValue, newsz);
+    }
+
+    void Write(const uint8_t * buffer) { memcpy(mValue, buffer, GetSize(buffer)); }
```

Comment:
What guarantees GetSize(buffer) bytes fit in mValue?

Reply:
> Well I was hoping the protocol layer wouldn't allow a write of a larger value than the underlying size

The protocol layer will not allow a write of a size larger than the attribute metadata size.  Is the size of mValue guaranteed to be at least that big?